### PR TITLE
labels: faster Compare function when using -tags stringlabels 

### PR DESCRIPTION
**Pull request by bboreham**
Instead of unpacking every individual string, we skip to the point where there is a difference.

Added a benchmark to illustrate.

[benchmark results updated 2023-06-15T1513]
Before/after this PR with `-tags stringlabels`:
```
goos: linux
goarch: amd64
pkg: github.com/prometheus/prometheus/model/labels
cpu: Intel(R) Core(TM) i5-9400F CPU @ 2.90GHz
                                 │  before.txt  │             after.txt              │
                                 │    sec/op    │   sec/op     vs base               │
Labels_Compare/equal-4              71.13n ± 6%   13.40n ± 2%  -81.16% (p=0.002 n=6)
Labels_Compare/not_equal-4          69.54n ± 2%   33.32n ± 2%  -52.09% (p=0.002 n=6)
Labels_Compare/different_sizes-4   39.050n ± 2%   9.840n ± 4%  -74.80% (p=0.002 n=6)
Labels_Compare/lots-4              300.85n ± 9%   46.59n ± 4%  -84.51% (p=0.002 n=6)
geomean                             87.31n        21.27n       -75.64%
```

After this PR, without/with `-tags stringlabels`:
```
goos: linux
goarch: amd64
pkg: github.com/prometheus/prometheus/model/labels
cpu: Intel(R) Core(TM) i5-9400F CPU @ 2.90GHz
                                 │  slice.txt   │             after.txt              │
                                 │    sec/op    │   sec/op     vs base               │
Labels_Compare/equal-4              23.25n ± 3%   13.40n ± 2%  -42.35% (p=0.002 n=6)
Labels_Compare/not_equal-4          25.02n ± 7%   33.32n ± 2%  +33.20% (p=0.002 n=6)
Labels_Compare/different_sizes-4   14.365n ± 6%   9.840n ± 4%  -31.50% (p=0.002 n=6)
Labels_Compare/lots-4               92.77n ± 3%   46.59n ± 4%  -49.77% (p=0.002 n=6)
geomean                             29.67n        21.27n       -28.31%
```

### CODE REVIEW

File: model/labels/labels_stringlabels.go
Comment (colega) [inline model/labels/labels_stringlabels.go:452]: Not sure about this, I can see the readability improvement in your version with dedidated `else if` clause, but it leaves the doubt of whether there's an option to not return from  `if firstCharDifferent == len(shorter)`.
Maybe an explicit return and also a comment?

```suggestion
		}
		// b is shorter.
		return +1
```

File: model/labels/labels_stringlabels.go
Comment (colega) [inline model/labels/labels_stringlabels.go:442]: Since `decodeSize` is very cheap, we can keep decoding sizes as we go, like:
```suggestion
	i := 0
	lastLblStart, nextLblStart := 0, 0
	_ = longer[len(shorter)-1] // Get compiler to do bounds-check on longer just once here.
	for ; i < len(shorter); i++ {
		if shorter[i] != longer[i] {
			break
		}
		if i >= nextLblStart {
			size, nextI := decodeSize(shorter, i)
			nextLblStart = nextI + size
			lastLblStart = i
		}
	}
```

So then later we can just do `i = lastLblStart`.

This gives pretty good performance impact:
```
goos: linux
goarch: amd64
pkg: github.com/prometheus/prometheus/model/labels
cpu: 11th Gen Intel(R) Core(TM) i7-11700K @ 3.60GHz
                                  │ bryanagain  │              tracknoloop               │
                                  │   sec/op    │   sec/op     vs base                   │
Labels_Compare/equal-16             25.59n ± 0%   26.48n ± 3%   +3.46% (p=0.000 n=30+10)
Labels_Compare/not_equal-16         30.25n ± 1%   24.07n ± 3%  -20.43% (p=0.000 n=30+10)
Labels_Compare/different_sizes-16   16.46n ± 1%   12.38n ± 8%  -24.84% (p=0.000 n=30+10)
Labels_Compare/lots-16              47.77n ± 1%   42.97n ± 4%  -10.06% (p=0.000 n=30+10)
geomean                             27.94n        24.13n       -13.63%

```

I've ran this benchmark multiple times and didn't get a consistent result (I was getting the `equals` benchmark faster in my version, which doesn't make sense), however I also think that if we do this, we can remove the last for loop (as we already know which labels are different, no need to search for them).

Comment (colega) [reply model/labels/labels_stringlabels.go:442]: In other words, I think we can do this later:
```go
	aStr, _ := decodeString(a.data, lastLblStart)
	bStr, _ := decodeString(b.data, lastLblStart)
```

File: model/labels/labels_stringlabels.go
Comment (bboreham) [reply model/labels/labels_stringlabels.go:449]: I read what I wrote as "if the first different character is at the end of `a.data`".

I think yours requires to be thinking about longer and shorter, which I'm not at this point.

File: model/labels/labels_stringlabels.go
Comment (bboreham) [reply model/labels/labels_stringlabels.go:444]: I would like to search for the first difference much faster, e.g. 8 bytes at a time using dwords, or even longer strides using SIMD.  I made various attempts at this, e.g. 06f8bc7787435a93626c96c3f0ea53b724ef23b4.  I think it's faster than your version, although I don't really like the implementation in `github.com/grailbio/base/simd.FirstUnequal8Unsafe`.

Putting an extra if-statement inside the tight loop does not seem attractive.

File: model/labels/labels_stringlabels.go
Comment (colega) [reply model/labels/labels_stringlabels.go:449]: I understand from `if firstCharDifferent == len(shorter) {` that `firstCharDifferent` actually lost its value, as it's not _first char different_ now, but a signal saying: _ all common characters are equal, pending to check whether one string is  longer than the other one._

Which now makes me think that it would be better  to just write:

```go
	if firstCharDifferent == len(shorter) {
		return compareLengths(a.data, b.data)
	}
	// ...
}
// ...

func compareLengths(a, b []string) int {
	switch {
	case len(a) == len(b):
		return 0
	case len(a) < len(b):
		return -1
	case len(b) > len(a):
		return +1
	}
}
```

File: model/labels/labels_stringlabels.go
Comment (bwplotka) [inline model/labels/labels_stringlabels.go:430]: No need for `else` - it can be standalone if (readability nit, I don't expect this to change efficiency)

File: model/labels/labels_stringlabels.go
Comment (bwplotka) [inline model/labels/labels_stringlabels.go:449]:
```suggestion
		}
		if firstCharDifferent == len(a.data) {
```

File: model/labels/labels_stringlabels.go
Comment (bwplotka) [reply model/labels/labels_stringlabels.go:449]: Both options are fine to me, but I agree with @colega naming of `firstCharDifferent` is indeed misleading (premature - it might be just all equal characters. Perhaps keeping `i` and not introducing new var might help here?

**Review by bwplotka — APPROVED** [review]
Looks good 💪🏽  Thanks!

File: model/labels/labels_stringlabels.go
Comment (bboreham) [reply model/labels/labels_stringlabels.go]: I included your latest code, and updated my benchmark results.
Now mostly going faster than the slice version.

File: model/labels/labels_stringlabels.go
Comment (bboreham) [reply model/labels/labels_stringlabels.go]: Done - added #12455 and I don't think nonalphabetic sorting is achieveable.

**Review by colega — APPROVED** [review]
:rocket:

**Comment by bboreham** [comment]
@bwplotka a couple of things changed since you last approved so I'll let you take another look before merging.

**Review by SuperQ — APPROVED** [review]
:shipit:

**Pull request merged**
